designs/bare-metal-out-of-band-webhook.md

```diff
@@ -0,0 +1,549 @@
+
+Rufio currently requires a secret for [each machine](https://github.com/tinkerbell/rufio/blob/54081cd4ecad76e7a6643ec804982ff279bad0e1/controllers/task.go#L93). The webhook provider will use the same secret(s) for HMAC signing for all machines. Right now, we create all secrets via EKS Anywhere CLI. This pattern will not change. We will be updating the EKS Anywhere code base to use the same secret(s) for all machines.
+
+#### Payload
+
+The core of the webhook provider will be constructing an HTTP request to send to the customer defined URL. The webhook provider will send notifications to a single endpoint that must accept the same payload spec for all BMC request (power on, power off, etc). Webhook consumers will be required to parse the payload to determine which BMC action is being requested.
+
+Payload example with all fields (in operation, EKS Anywhere will only send single task requests).
+
+```json
+{
+  "host": "192.168.2.3",
+  "task": {
+    "power": "on",
+    "bootDevice": {
+      "device": "pxe",
+      "persistent": true,
+      "efiBoot": true
+    },
+    "virtualMedia": {
+      "mediaUrl": "https://boot.netboot.xyz/ipxe/netboot.xyz.iso",
+      "kind": "CD"
+    }
+  }
+}
+```
+
+Example payload for setting the next boot device to “pxe”.
+
+```json
+{
+   "host": "192.168.2.2",
+   "task": {
+        "bootDevice": {
+            "device": "pxe",
+            "persistent": false,
+            "efiBoot": false
+        }
+   }
+}
+```
+
+#### OpenAPI v3 Spec
+
+The following OpenAPI v3 spec is an example of what we will provide customers to aid in any development of a webhook consumer they might do. This is just an example. This spec can be put into an [online openapi editor](https://editor.swagger.io/) for manipulation and for generating  clients and servers. It is important to say again that EKS Anywhere will not be implementing any server functionality. Only client functionality.
+
+```yaml
+openapi: 3.0.3
+info:
+  title: Webhook Receiver - OpenAPI 3.0
+  description: |-
+    Webhook notification spec
+  version: "0.1"
+servers:
+  - url: https://localhost:9000
+tags:
+  - name: BMC
+    description: Webhook listener notification for BMC interactions
+paths:
+  /webhook: # This URI is just an example as it is completely defined by the customer
+    post:
+      tags:
+        - BMC
+      summary: webhook listener for BMC calls
+      description: All BMC actions should be handle here. The payload will be different
+      requestBody:
+        description: Webhook for all BMC calls
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/BMC'
+        required: true
+      responses:
+        '200':
+          description: Successful operation
+components:
+  schemas:
+    BMC:
+      type: object
+      properties:
+        host:
+          type: string
+          example: "192.168.2.3"
+          description: The IP or hostname of a BMC
```

Comment:
Are we confident customers are content documenting address information of their BMCs like this? Perhaps opaque identifiers could be used instead (which makes them webhook provider implementation dependent so they could still use hostnames etc) with the webhook providers dealing with conversion how they see fit.

Reply:
yeah, the data type is a string and we wont do any validation around it so a customer could definitely use an opaque identifier instead of an IP. I will make a note of this so that we get docs around this.